File: packages/shared/src/types/domain.ts
import type { ActivityLevel, Goal, MealType, TargetDuration, UnitSystem } from '../constants.js'

export interface GoalRecord {
  id: string
  goalType: Goal
  startingWeightKg: number
  targetWeightKg: number
  targetDuration: TargetDuration
  customEndDate: string | null
  activityLevel: ActivityLevel
  bmr: number
  tdee: number
  calorieTarget: number
  calorieTargetOverridden: boolean
  proteinTargetG: number
  proteinTargetOverridden: boolean
  carbsTargetG: number
  carbsTargetOverridden: boolean
  fatTargetG: number
  fatTargetOverridden: boolean
  fibreTargetG: number
  fibreTargetOverridden: boolean
  waterTargetMl: number
  waterTargetOverridden: boolean
  isActive: boolean
  createdAt: string
  updatedAt: string
}

export interface FoodRecord {
  id: string
  source: 'openfoodfacts' | 'usda' | 'custom'
  sourceId: string | null
  barcode: string | null
  name: string
  brand: string | null
  servingSize: number
  servingUnit: string
  calories: number
  proteinG: number
  carbsG: number
  fatG: number
  fibreG: number
  isFavourite?: boolean
  /**
   * True if this food already has a durable row in D1 (custom, or a
   * provider result someone has actually logged/favourited before).
   * False for a live external-provider search hit that only exists in
   * memory for this response — logging or favouriting it materializes a
   * real row first (see foods/external).
   */
  isLocal: boolean
}

export interface RecognizedFoodItem {
  name: string
  estimatedQuantityGrams: number
  /** 0-1 — the AI's own confidence in this identification, not a nutrition-data confidence. */
  confidence: number
  /** The best nutrition-database match for `name`, if any — AI never supplies nutrition values itself. */
  matchedFood: FoodRecord | null
}

export interface PhotoAnalysisResult {
  imageHash: string
  /** Pass this straight through as the meal's `photoR2Key` when saving. */
  r2Key: string
  items: RecognizedFoodItem[]
  /** True if this exact photo (by content hash) was already analyzed before — no new AI call was made. */
  fromCache: boolean
}

/**
 * Fields read directly off a nutrition facts label photo — null wherever the
 * label was illegible or didn't print that field. Pre-fills the custom-food
 * form for the user to confirm/correct, same as CreateCustomFoodInput's shape.
 */
export interface LabelAnalysisResult {
  name: string | null
  brand: string | null
  servingSize: number | null
  servingUnit: string | null
  calories: number | null
  proteinG: number | null
  carbsG: number | null
  fatG: number | null
  fibreG: number | null
}

export interface RecipeItemRecord {
  id: string
  foodId: string
  food?: FoodRecord
  quantity: number
  unit: string
}

export interface RecipeRecord {
  id: string
  name: string
  servings: number
  instructions: string | null
  items: RecipeItemRecord[]
  /** Per-serving totals, derived from items and `servings`. */
  caloriesPerServing: number
  proteinGPerServing: number
  carbsGPerServing: number
  fatGPerServing: number
  fibreGPerServing: number
  createdAt: string
  updatedAt: string
}

export interface MealItemRecord {
  id: string
  foodId: string | null
  recipeId: string | null
  food?: FoodRecord
  quantity: number
  unit: string
  calories: number
  proteinG: number
  carbsG: number
  fatG: number
  fibreG: number
  aiConfidence: number | null
}

export interface MealRecord {
  id: string
  mealType: MealType
  loggedAt: string
  photoR2Key: string | null
  notes: string | null
  totalCalories: number
  totalProteinG: number
  totalCarbsG: number
  totalFatG: number
  totalFibreG: number
  items: MealItemRecord[]
  createdAt: string
  updatedAt: string
}

export interface WaterLogRecord {
  id: string
  amountMl: number
  loggedAt: string
  createdAt: string
}

export interface WeightLogRecord {
  id: string
  weightKg: number
  loggedAt: string
  notes: string | null
  createdAt: string
}

export interface SettingsRecord {
  unitSystem: UnitSystem
  theme: 'light' | 'dark' | 'system'
  updatedAt: string
}

export interface NotificationSettingsRecord {
  onesignalPlayerId: string | null
  breakfastReminderTime: string | null
  lunchReminderTime: string | null
  dinnerReminderTime: string | null
  waterReminderEnabled: boolean
  waterReminderIntervalMinutes: number | null
  weighInReminderTime: string | null
  weighInReminderDays: number[]
  quietHoursStart: string | null
  quietHoursEnd: string | null
  timezone: string
  updatedAt: string
}

export interface CustomReminderRecord {
  id: string
  label: string
  time: string
  daysOfWeek: number[]
  enabled: boolean
  createdAt: string
}

File: apps/ios/OmNomNom/Models/Domain.swift
import Foundation

// Ported from packages/shared/src/types/domain.ts + types/user.ts.
// Date-ish fields are kept as ISO strings to match the API exactly; use
// `ISO8601.date(from:)` where a Date is needed.

struct PublicUser: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var name: String
    var email: String
    var dateOfBirth: String        // YYYY-MM-DD
    var sex: Sex
    var heightCm: Double
    var createdAt: String
    var updatedAt: String
}

struct GoalRecord: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var goalType: Goal
    var startingWeightKg: Double
    var targetWeightKg: Double
    var targetDuration: TargetDuration
    var customEndDate: String?
    var activityLevel: ActivityLevel
    var bmr: Double
    var tdee: Double
    var calorieTarget: Double
    var calorieTargetOverridden: Bool
    var proteinTargetG: Double
    var proteinTargetOverridden: Bool
    var carbsTargetG: Double
    var carbsTargetOverridden: Bool
    var fatTargetG: Double
    var fatTargetOverridden: Bool
    var fibreTargetG: Double
    var fibreTargetOverridden: Bool
    var waterTargetMl: Double
    var waterTargetOverridden: Bool
    var isActive: Bool
    var createdAt: String
    var updatedAt: String
}

enum FoodSource: String, Codable, Hashable, Sendable {
    case openfoodfacts, usda, custom
}

struct FoodRecord: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var source: FoodSource
    var sourceId: String?
    var barcode: String?
    var name: String
    var brand: String?
    var servingSize: Double
    var servingUnit: String
    var calories: Double
    var proteinG: Double
    var carbsG: Double
    var fatG: Double
    var fibreG: Double
    var isFavourite: Bool?
    /// false = a live external search hit not yet in the DB. Must be materialized
    /// via POST /api/foods/external before logging or favouriting.
    var isLocal: Bool
}

struct RecognizedFoodItem: Codable, Hashable, Sendable {
    var name: String
    var estimatedQuantityGrams: Double
    var confidence: Double
    var matchedFood: FoodRecord?
}

struct PhotoAnalysisResult: Codable, Sendable {
    var imageHash: String
    var r2Key: String
    var items: [RecognizedFoodItem]
    var fromCache: Bool
}

struct LabelAnalysisResult: Codable, Sendable {
    var name: String?
    var brand: String?
    var servingSize: Double?
    var servingUnit: String?
    var calories: Double?
    var proteinG: Double?
    var carbsG: Double?
    var fatG: Double?
    var fibreG: Double?
}

struct RecipeItemRecord: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var foodId: String
    var food: FoodRecord?
    var quantity: Double
    var unit: String
}

struct RecipeRecord: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var name: String
    var servings: Double
    var instructions: String?
    var items: [RecipeItemRecord]
    var caloriesPerServing: Double
    var proteinGPerServing: Double
    var carbsGPerServing: Double
    var fatGPerServing: Double
    var fibreGPerServing: Double
    var createdAt: String
    var updatedAt: String
}

struct MealItemRecord: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var foodId: String?
    var recipeId: String?
    var food: FoodRecord?
    var quantity: Double
    var unit: String
    var calories: Double
    var proteinG: Double
    var carbsG: Double
    var fatG: Double
    var fibreG: Double
    var aiConfidence: Double?
}

struct MealRecord: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var mealType: MealType
    var loggedAt: String
    var photoR2Key: String?
    var notes: String?
    var totalCalories: Double
    var totalProteinG: Double
    var totalCarbsG: Double
    var totalFatG: Double
    var totalFibreG: Double
    var items: [MealItemRecord]
    var createdAt: String
    var updatedAt: String
}

struct WaterLogRecord: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var amountMl: Double
    var loggedAt: String
    var createdAt: String
}

struct WeightLogRecord: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var weightKg: Double
    var loggedAt: String
    var notes: String?
    var createdAt: String
}

struct SettingsRecord: Codable, Hashable, Sendable {
    var unitSystem: UnitSystem
    var theme: AppTheme
    var updatedAt: String
}

struct NotificationSettingsRecord: Codable, Hashable, Sendable {
    var onesignalPlayerId: String?
    var breakfastReminderTime: String?
    var lunchReminderTime: String?
    var dinnerReminderTime: String?
    var waterReminderEnabled: Bool
    var waterReminderIntervalMinutes: Int?
    var weighInReminderTime: String?
    var weighInReminderDays: [Int]
    var quietHoursStart: String?
    var quietHoursEnd: String?
    var timezone: String
    var updatedAt: String
}

struct CustomReminderRecord: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var label: String
    var time: String
    var daysOfWeek: [Int]
    var enabled: Bool
    var createdAt: String
}

struct PublicWidgetToken: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var label: String
    var lastUsedAt: String?
    var createdAt: String
}

struct IssuedWidgetToken: Codable, Identifiable, Hashable, Sendable {
    let id: String
    var label: String
    var lastUsedAt: String?
    var createdAt: String
    var token: String
}

// MARK: - Nutrition score

struct ScoreComponent: Codable, Hashable, Sendable {
    var points: Double
    var maxPoints: Double
}

struct NutritionScoreComponents: Codable, Hashable, Sendable {
    var calories: ScoreComponent
    var protein: ScoreComponent
    var fibre: ScoreComponent
    var foodQuality: ScoreComponent
    var consistency: ScoreComponent
    var water: ScoreComponent
    var loggingCompleteness: ScoreComponent
}

struct NutritionScoreBreakdown: Codable, Hashable, Sendable {
    var score: Double
    var label: NutritionScoreLabel
    var components: NutritionScoreComponents
}

struct DailyScoreSummary: Codable, Identifiable, Hashable, Sendable {
    var dateKey: String
    var score: Double
    var label: NutritionScoreLabel
    var id: String { dateKey }
}

// MARK: - Widget summary

struct WidgetMacro: Codable, Hashable, Sendable {
    var consumed: Double
    var target: Double
}

struct WidgetSummary: Codable, Hashable, Sendable {
    enum MascotMood: String, Codable, Sendable {
        case excited, happy, neutral, sad
    }
    var score: Double
    var label: NutritionScoreLabel
    var message: String
    var mascotMood: MascotMood
    var macros: Macros
    struct Macros: Codable, Hashable, Sendable {
        var calories: WidgetMacro
        var protein: WidgetMacro
        var fat: WidgetMacro
    }
}
